tracks/python/exercises/isbn-verifier/mentoring.md

```diff
@@ -0,0 +1,111 @@
+
+## Problem and Challenges
+
+The problem asks you to write an ISBN validation checker.
+
+* Any "-" in the input string should be ignored.
+* A valid ISBN must be 10 values long.
+* A valid ISBN must be entirely digits, other than the last character.
+* A valid ISBN may have `"X"` as the last character, which should be treated as a `10` value.
+* The ISBN is valid if `sum(value * position_counter) % 11 == 0` for `value in <input string>` and `position_counter in <(10..1)>`.
+
+## Reasonable Solution
+
+```python
+def is_valid(isbn: str) -> bool:
+    values = [value for value in isbn if value != "-"]
+
+    if len(values) != 10:
+        return False
+
+    if values[-1] == "X":
+        values[-1] = "10"
+
+    if any(not value.isdigit() for value in values):
+        return False
+
+    total = sum(int(value) * (10 - position) for position, value in enumerate(values))
+    return total % 11 == 0
+```
+
+## Talking Points
+
+### Generator expressions
+
+The sum can be computed using a `for` loop or via the `sum()` [built-in function][builtins-functions].
+
+```python
+total = 0
+for position, value in enumerate(values)):
+    total += int(value) * (10 - position)
+```
+
+Generator expressions are considered very Pythonic and enable compact and clean iteration.
+They are also very efficient.
+
+### The position value
+
+The position value can be directly generated using `range()`.
+However, using `range()` requires mapping the position value to an ISBN positional value, by computing an index or via `zip()`.
+
+```python
+total = 0
+for position, value in zip(range(10, 0, -1), values):
+    total += int(value) * position
+```
+
+The [`enumerate()` function][builtins-functions] provides a nice way to get both a position and a value.
+The position returned by `enumerate()` does need to be modified (`10 - position`) to be used.
+
+### The trailing "X" value
+
+The final value can be handled in the summation loop by checking the position and the value and replacing "X" with 10.
+Doing so allows for keeping the ISBN as a `str` instead of converting it to `list[str]` to be able to hold a `"10"` value.
+
+```python
+total = 0
+for position, value in enumerate(isbn):
+    if position == 9 and value == "X":
+        total += 10
+    else:
+        total += int(value) * (10 - position)
+```
+
+Recognizing that changing the data from `str` to `list[str]` allows for more flexibility of what it can store (`"10"`) and a simpler summation loop.
+
+### Checking for non-digits
+
+The [built-in `any()` function][builtin-functions] is very convenient for any non-digit values in the input.
+If the input is first converted to a string and a trailing "X" is replaced with a digit, the digit check is fairly straight forward.
+If the trailing "X" is not handled prior to the digit check, the final character needs to be handled separately.
+
+```python
+if any(not value.isdigit() for value in isbn[:-1]):
+    return False
+if not isbn[-1].isdigit() and isbn[-1] != "X":
```

Comment:
I tend to shy away from a double negative in the conditional statement, is there another way to express this with out the not and not equal?

Reply:
This isn't a double negative as the two negatives are applying to different conditionals. Do you prefer this?

```suggestion
if isbn[-1].isdigit() or isbn[-1] = "X":
    pass
else:
```

Reply:
Yes, I prefer that, it is a more positive thing.  It also has the different body, from `return False` to just a `pass`.  Surprising to me, in a way, as I would have thought the other potential would have been `True`.  Because with the else, that should be the true branch, and I prefer true and then false conditional branches to return true and then false.  Otherwise I usually consider the conditional statement to be "backward".

Reply:
This is one of the many tests required to check if a string is a valid ISBN. All the tests need to pass before we can `return True`. I would argue that

```
if not condition:
    return False
```
is significantly more compact, readable and Pythonic than:
```
if condition:
    pass
else:
    return False
```

What if I wrapped it all in a single `not`?

```suggestion
if not (isbn[-1].isdigit() or isbn[-1] == "X"):
```

Reply:
With the following line being `return False` I do question if that is correct, as the last "digit" is allowed to be "X".

It is part of the larger `any()` statement, so I will defer to you for the specifics here.  It is only that I wanted to consider options here.

Reply:
This line is separate from the `any()` statement. The last character of the ISBN is allowed to be `"X"`, per the problem spec.